File: src/llm.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Optional, List, Dict, Any
import os
import json
import requests

@dataclass
class LLMConfig:
    provider: str = "openai"   # "openai" або "ollama"
    model: str = "gpt-4o-mini" # або "llama3.1"
    api_key: str = ""          # для openai
    base_url: str = ""         # для ollama (http://localhost:11434)

SYSTEM_PROMPT = """Ти — BioConsult, навчальний консультант з біології.
Мова: українська. Стиль: чітко, структуровано, без зайвої води.

ГОЛОВНЕ ПРАВИЛО:
- Якщо надано контекст (RAG) — опирайся на нього як на основне джерело.
- Якщо контексту недостатньо — прямо скажи, чого бракує, і попроси додати матеріали.

ФОРМАТ:
1) Коротка відповідь (1–3 речення)
2) Пояснення списком
3) (Опційно) приклад/аналогія
4) Джерела: [#1], [#2] лише якщо реально використав RAG.
"""

def _format_context(contexts: List[Dict[str, Any]]) -> str:
    if not contexts:
        return ""
    blocks = []
    for i, c in enumerate(contexts, start=1):
        src = c.get("meta", {}).get("source", "джерело")
        txt = (c.get("text") or "").strip()
        blocks.append(f"[#{i} {src}] {txt}")
    return "\n\n".join(blocks)

class LLM:
    def __init__(self, cfg: LLMConfig):
        self.cfg = cfg

    def generate(self, user_text: str, contexts: List[Dict[str, Any]]) -> str:
        ctx = _format_context(contexts)

        if self.cfg.provider == "openai":
            return self._openai(user_text, ctx)
        if self.cfg.provider == "ollama":
            return self._ollama(user_text, ctx)

        raise ValueError(f"Unknown provider: {self.cfg.provider}")

    def _openai(self, user_text: str, ctx: str) -> str:
        # OpenAI Responses API (сучасна)
        import requests

        api_key = self.cfg.api_key or os.getenv("OPENAI_API_KEY", "")
        if not api_key:
            raise RuntimeError("OPENAI_API_KEY is missing")

        content = user_text
        if ctx:
            content += "\n\nКонтекст (RAG):\n" + ctx

        payload = {
            "model": self.cfg.model,
            "input": [
                {"role": "system", "content": [{"type": "text", "text": SYSTEM_PROMPT}]},
                {"role": "user", "content": [{"type": "text", "text": content}]}
            ]
        }

        r = requests.post(
            "https://api.openai.com/v1/responses",
            headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
            json=payload,
            timeout=60
        )
        if not r.ok:
            raise RuntimeError(r.text)

        data = r.json()
        # частий випадок: output_text
        if isinstance(data.get("output_text"), str) and data["output_text"]:
            return data["output_text"]

        # fallback: витяг з output[]
        out = data.get("output", [])
        for item in out:
            for c in item.get("content", []):
                if c.get("type") in ("output_text", "text") and isinstance(c.get("text"), str):
                    return c["text"]

        return ""

    def _ollama(self, user_text: str, ctx: str) -> str:
        base = self.cfg.base_url or os.getenv("OLLAMA_BASE_URL", "http://localhost:11434")
        prompt = SYSTEM_PROMPT + "\n\n"
        if ctx:
            prompt += "Контекст (RAG):\n" + ctx + "\n\n"
        prompt += "Запит користувача:\n" + user_text

        # Ollama generate API
        url = f"{base}/api/generate"
        payload = {"model": self.cfg.model, "prompt": prompt, "stream": False}

        r = requests.post(url, json=payload, timeout=120)
        if not r.ok:
            raise RuntimeError(r.text)
        return (r.json().get("response") or "").strip()

File: src/llm.js
function buildSystem(){
  return [
    "Ти — BioConsult, консультант з біології.",
    "Відповідай українською, просто і точно.",
    "Якщо є RAG-контекст — використовуй його в першу чергу.",
    "Якщо даних недостатньо — скажи, що саме потрібно уточнити.",
    "Додай короткий блок 'Джерела' з позначками [#1], [#2] (тільки якщо використовував контекст)."
  ].join("\n");
}

function buildContextBlock(contexts){
  if(!contexts?.length) return "";
  return contexts.map((c, i) => `[#${i+1} ${c.title}] ${c.text}`).join("\n\n");
}

function sourcesFromContexts(contexts){
  return (contexts || []).map(c => ({
    title: c.title,
    snippet: (c.text || "").slice(0, 200) + ((c.text || "").length > 200 ? "…" : "")
  }));
}

function extractOutputText(data){
  if (typeof data?.output_text === "string" && data.output_text) return data.output_text;

  const out = data?.output;
  if (Array.isArray(out)) {
    for (const item of out) {
      const content = item?.content;
      if (Array.isArray(content)) {
        for (const c of content) {
          if (c?.type === "output_text" && typeof c?.text === "string") return c.text;
          if (c?.type === "text" && typeof c?.text === "string") return c.text;
        }
      }
    }
  }
  return "";
}

export const LLM = {
  async answer({ apiKey, model, userText, contexts, image }){
    const system = buildSystem();
    const ctx = buildContextBlock(contexts);

    const userParts = [{ type:"text", text: userText }];

    if (image?.url) {
      userParts.push({ type:"text", text: `\n(Додано зображення: ${image.label || "image"})\n` });
      userParts.push({ type:"image_url", image_url: { url: image.url } });
    }

    if (ctx) userParts.push({ type:"text", text: `\n\nКонтекст (RAG):\n${ctx}` });

    const body = {
      model: model || "gpt-4o-mini",
      input: [
        { role:"system", content:[{ type:"text", text: system }] },
        { role:"user", content: userParts }
      ]
    };

    const res = await fetch("https://api.openai.com/v1/responses", {
      method:"POST",
      headers:{
        "Content-Type":"application/json",
        "Authorization":"Bearer " + apiKey
      },
      body: JSON.stringify(body)
    });

    if(!res.ok){
      const t = await res.text();
      throw new Error(t || ("HTTP " + res.status));
    }

    const data = await res.json();
    const text = extractOutputText(data) || "(порожня відповідь)";

    return { answer: text, sources: sourcesFromContexts(contexts) };
  }
};
